tempesta_fw/ss_skb.c

```diff
@@ -976,3 +976,105 @@ ss_skb_split(struct sk_buff *skb, int len)
 
 	return buff;
 }
+
+static inline int
+__coalesce_frag(SsSkbList *skb_list, skb_frag_t *frag)
+{
+	struct sk_buff *skb = ss_skb_peek_tail(skb_list);
+
+	if (!skb || skb_shinfo(skb)->nr_frags == MAX_SKB_FRAGS) {
+		skb = ss_skb_alloc();
+		if (!skb)
+			return -ENOMEM;
+		ss_skb_queue_tail(skb_list, skb);
+	}
+
+	skb_shinfo(skb)->frags[skb_shinfo(skb)->nr_frags++] = *frag;
+	ss_skb_adjust_data_len(skb, frag->size);
+	__skb_frag_ref(frag);
+
+	return 0;
+}
+
+static int
+ss_skb_queue_coalesce_tail(SsSkbList *skb_list, const struct sk_buff *skb)
+{
+	int i;
+	skb_frag_t head_frag;
+	unsigned int headlen = skb_headlen(skb);
+
+	if (headlen) {
+		BUG_ON(!skb->head_frag);
+		head_frag.size = headlen;
+		head_frag.page.p = virt_to_head_page(skb->head);
+		head_frag.page_offset = skb->data -
+			(unsigned char *)page_address(head_frag.page.p);
+		if (__coalesce_frag(skb_list, &head_frag))
+			return -ENOMEM;
+	}
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		if (__coalesce_frag(skb_list, &skb_shinfo(skb)->frags[i]))
```

Comment:
I think this makes things unnecessary complex. Technically `ss_skb_unroll()` does only following things:

1. if `skb` has head, then shift frags to the left for one position and map `skb->head` page to `skb->frags[0]`;

2. if `skb` has flag_list, then move the skbs to `skb_list`.

So in some cases you can just do plain `memmove()` on `skb->frags` instead of calling `__coalesce_frag()` in the loop. Moreover, if `skb->frags` is full, then you still can do `memmove()` for `nr_frags - 1` items and try to place the last frag in first skb fragment from `frag_list` or, at the end, allocate a new skb.

I think the code can be shorter and more efficient.